### PR TITLE
ws: Disallow direct URL logins with LoginTo=false

### LINKED ISSUE #15077: unauthenticated server-side request forgery - login page allows ssh to to localhost [CVE-2020-35850]
Cockpit version: 234
OS: Ubuntu 18.04
Page: login

User can detect open ssh port or another open ports on server that services Cockpit last version. This is a vulnerability that allows an user send request to internal hosts for detecting open ports. So that firewall configuration can be bypassed or the server can be used like gateway by malicious user. 
In addition, user induces the application to make an request back to the server that is hosting Cockpit. 
For example: if system admin creates iptables rule to drop all packets that come to 22 port or another port, user can detect whether port 22 is open or not.

Assuming that there is a iptables rule which port 22 is open for 127.0.0.1 (loopback interface) but is closed for other interfaces 

```
First HTTP Request:

GET /cockpit+=192.168.1.27:22/login HTTP/1.1
Host: 192.168.1.27:9090
User-Agent: *
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Authorization: Basic dWJ1bnR1OnVidW50dQ==
X-Authorize:
Connection: close
Cookie: cockpit=deleted
```
```
Second HTTP Request:

GET /cockpit+=127.0.0.1:22/login HTTP/1.1
Host: 192.168.1.27:9090
User-Agent: *
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Authorization: Basic dWJ1bnR1OnVidW50dQ==
X-Authorize:
Connection: close
Cookie: cockpit=deleted
```

Steps to reproduce:
On login panel,
1-Click Other Options.
2-Set 127.0.0.1 to Connect to field and send request with incorrect credentials.
3-Intercept the request with Burp Suite
4- If ssh service is open on port 22 and credentials are wrong, server returns "401 Authentication Failed" response.
5-If user tries connect to a port that accepts data for ssh connection , server returns "401 Authentication failed: no-host" response and waits 10 seconds.
6- If user tries connect to a closed port , server returns "401 Authentication failed: no-host" response and without waiting.
7- If ssh service is open on port 22 and credentials are correct, server returns "200" response:

[PoC video 1](https://github.com/passtheticket/vulnerability-research/raw/main/cockpitProject/1.mp4)
[PoC video 2](https://github.com/passtheticket/vulnerability-research/raw/main/cockpitProject/2.mp4)
[PoC video 3](https://github.com/passtheticket/vulnerability-research/raw/main/cockpitProject/3.mp4)

### LINKED ISSUE #18540: cockpit: LoginTo=false does not prevent connections to remote hosts
### Explain what happens

In issue #17340, the [WebService > LoginTo](https://cockpit-project.org/guide/latest/cockpit.conf.5.html) option was pointed out as a way to disable the "Other Options > Connect to" option on the Cockpit login screen.  In the man page, that option is documented as:

> When set to true the Connect to option on the login screen is visible and allows logging into another server. If this option is not specified then it will be automatically detected based on whether the cockpit-ssh process is available or not.

That documentation is not explicit what is the expected behaviour in case of LoginTo=false - whether it's only supposed to hide the option, or actually also prevent logging to remote hosts.  The latter is likely to be assumed.  However, the option only hides the input field on the login screen without preventing remote logins.

To test, set the following in the cockpit.conf:

```
[WebService]
LoginTo=false
```

Restart the service and visit the login page.  See that the "Connect to" option is not shown.  Change the URL of the page to append `/=other.host.or.ip`.  I.e the URL as:

https://cockpit.example.com:9090/

is changed to:

https://cockpit.example.com:9090/=otherhost.example.com/

or

https://cockpit.example.com:9090/=10.0.0.10/

See that after loading such changed URL, the login page now indicates the remote host as the "Server".  Providing correct authentication credentials for the remote hosts opens a session on that remote host.

### Version of Cockpit

Tested with RHEL packaged Cockpit 276.1

### Where is the problem in Cockpit?

None

### Server operating system

None

### Server operating system version

_No response_

### What browsers are you using?

_No response_

### System log

_No response_

### PR DESCRIPTION
**Pull request by martinpitt**
The current documentation of LoginTo= isn't very specific about what exactly happens with a "false" value; but it is plausible for an admin to assume that "false" would disallow logging into a remote host completely -- not merely hide the "Connect to:" field and then allowing a direct URL login anyway.

It is sometimes important to disallow direct SSH logins from the login page on publicly exposed bastion hosts, as this functionality allows unauthenticated remote users to:

 - scan the internal network for existing hosts, which might otherwise not be accessible directly from the internet (Fixes #18540, https://bugzilla.redhat.com/show_bug.cgi?id=2167006)

 - scan the cockpit-ws host or internal network hosts for open ports (Fixes #15077, https://bugzilla.redhat.com/show_bug.cgi?id=2018741)

So change ws to reject direct URL logins with `LoginTo=false`. This happens most naturally in cockpit_session_launch(), as we still want to allow remote URLs from the shell's host switcher in already authenticated sessions. This will not produce a very friendly error message, but it doesn't have to be -- at that point specifying direct URLs can be considered hacking anyway.

Clarify the documentation accordingly.

### CODE REVIEW

File: doc/man/cockpit.conf.xml [hunk 1]
```diff
@@ -102,9 +102,15 @@ ForwardedForHeader = X-Forwarded-For
         <term><option>LoginTo</option></term>
         <listitem>
           <para>When set to <literal>true</literal> the <emphasis>Connect to</emphasis> option
-            on the login screen is visible and allows logging into another server. If this
-            option is not specified then it will be automatically detected based on whether
-            the <command>cockpit-ssh</command> process is available or not.</para>
+          on the login screen is visible and allows logging into another server. When set to
```
Comment (allisonkarlitskaya) [inline doc/man/cockpit.conf.xml:105]: The indentation change makes this patch hard to read.

I wonder if we should have the default just be False.  It's kinda dangerous, potentially unexpected, and `cockpit-ssh` is very easy to have installed.

It might break people's working configs on upgrade, but it would be easy to advise how to fix it.

File: src/ws/cockpitauth.c [hunk 2]
```diff
@@ -1047,6 +1047,13 @@ cockpit_session_launch (CockpitAuth *self,
       return NULL;
     }
 
+  /* this might be unset, which means "allow if cockpit-ssh is installed"; if it isn't, this will fail later on */
+  if (host && !cockpit_conf_bool ("WebService", "LoginTo", TRUE)) {
```
Comment (allisonkarlitskaya) [inline src/ws/cockpitauth.c:1051]: This looks sensible.

There are four cases where we could end up using cockpit-ssh.  Two of those involve explicit config and one of them involves a commandline option.  This should successfully neutralize the only case the user controls.

**Review by allisonkarlitskaya — DISMISSED** [review]

File: doc/man/cockpit.conf.xml (hunk 1 above)
Comment (martinpitt) [reply doc/man/cockpit.conf.xml:105]: Err right, that wasn't intended. vim's auto-spacing has an eternal fight with the weird indentation of our doc files. I fixed this.

Changing the default is a much harder decision. Connecting to a remote machine is a rather useful and important feature of cockpit, and just changing it on upgrades breaks existing setups too hard IMHO. Also, in some sense we even *want* people to do this "remote host" thing -- it's really not encouraged to put cockpit-ws on production servers, but always go through some bastion host. cockpit-desktop/Cockpit Client do exactly that (of course these can get an explicit config). `LoginTo` also interacts with `RequireHost=true` -- if that is set, then `LoginTo=false` does not make sense. So if it's unset, then it should be implicitly `true` with `RequireHost=true`.

In other words, this is a mine field -- at some point we need to analyze carefully what makes sense here, and probably clean up this non-orthogonal configuration.

File: src/ws/cockpitauth.c (hunk 2 above)
Comment (martinpitt) [reply src/ws/cockpitauth.c:1051]: So that's (1) login page with "Connect to:" (potentially dangerous), (2) shell's remote host (dangerous in a different way due to non-isolated frames, but not internal network scanning), (3) `cockpit-ws --local-session` (harmless). What is (4)?

**Review by allisonkarlitskaya — APPROVED** [review]
Reapproved after whitespace changes.
